src/plan.rs

```diff
@@ -0,0 +1,504 @@
+        Satisfier::check_older(self, s)
+    }
+
+    fn check_after(&self, l: LockTime) -> bool {
+        Satisfier::check_after(self, l)
+    }
+}
+
+/// Representation of a particular spending path on a descriptor. Contains the witness template
+/// and the timelocks needed for satisfying the plan.
+/// Calling `get_plan` on a Descriptor will return this structure,
+/// containing the cheapest spending path possible (considering the `Assets` given)
+#[derive(Debug, Clone)]
+pub struct Plan {
+    /// This plan's witness template
+    pub template: WitnessTemplate<Placeholder<DefiniteDescriptorKey>>,
+    /// The absolute timelock this plan uses
+    pub absolute_timelock: Option<LockTime>,
+    /// The relative timelock this plan uses
+    pub relative_timelock: Option<Sequence>,
+
+    pub(crate) desc_type: DescriptorType,
+}
+
+impl Plan {
+    /// Returns the witness version
+    pub fn witness_version(&self) -> Option<WitnessVersion> {
+        self.desc_type.segwit_version()
+    }
+
+    /// The weight, in witness units, needed for satisfying this plan (includes both
+    /// the script sig weight and the witness weight)
+    pub fn satisfaction_weight(&self) -> usize {
+        self.witness_size() + self.scriptsig_size() * 4
+    }
+
+    /// The size in bytes of the script sig that satisfies this plan
+    pub fn scriptsig_size(&self) -> usize {
+        match (self.desc_type.segwit_version(), self.desc_type) {
+            // Entire witness goes in the script_sig
+            (None, _) => witness_size(self.template.as_ref()),
+            // Taproot doesn't have a "wrapped" version (scriptSig len (1))
+            (Some(WitnessVersion::V1), _) => 1,
+            // scriptSig len (1) + OP_0 (1) + OP_PUSHBYTES_20 (1) + <pk hash> (20)
+            (_, DescriptorType::ShWpkh) => 1 + 1 + 1 + 20,
+            // scriptSig len (1) + OP_0 (1) + OP_PUSHBYTES_32 (1) + <script hash> (32)
+            (_, DescriptorType::ShWsh) | (_, DescriptorType::ShWshSortedMulti) => 1 + 1 + 1 + 32,
+            // Native Segwit v0 (scriptSig len (1))
+            __ => 1,
+        }
+    }
+
+    /// The size in bytes of the witness that satisfies this plan
+    pub fn witness_size(&self) -> usize {
+        if let Some(_) = self.desc_type.segwit_version() {
+            witness_size(self.template.as_ref())
+        } else {
+            0 // should be 1 if there's at least one segwit input in the tx, but that's out of
+              // scope as we can't possibly know that just by looking at the descriptor
+        }
+    }
+}
+
+/// The Assets we can use to satisfy a particular spending path
+#[derive(Debug, Default)]
+pub struct Assets {
+    keys: HashMap<hash160::Hash, DescriptorPublicKey>,
+    tap_key_spend_sig: Option<bitcoin::SchnorrSig>,
+    ecdsa_signatures: HashMap<DescriptorPublicKey, (bitcoin::EcdsaSig, usize)>,
+    schnorr_signatures: HashMap<(DescriptorPublicKey, TapLeafHash), (bitcoin::SchnorrSig, usize)>,
```

Comment:
Why do we need the signatures here? Will post a detailed reply in main conversation.

Reply:
The idea is that in multi-party protocols you can use signatures made by others as "assets" themselves. In that case you can't "cheat" by saying you have the key, because if you only have one signature you can either take it or it's not useful to you at all. When you say you have a key the planning code assumes you can sign anything you want with that key, which is a bit different.

Thinking about it now, I was planning to implement `IntoAssets` for a `psbt::Input` but I think I ended up forgetting about it, so I'll try to add that as well.

Reply:
I thought the purpose of planning was to decide which spend to use _before_ signing. In which case, why do we need signatures? I think having signatures in assets would be confusing, curious what @apoelstra thinks about this.

> In that case you can't "cheat" by saying you have the key

You would still need to assume that the other party will sign with a key because you calculated the change output based on it. So, they can still refuse to sign and cheat. 

Perhaps, we have some misunderstanding about what `Plan` is supposed to do.

Reply:
I think @afilini is thinking of the scenario of where you get a request to sign a TX  input and a few signatures under other keys. The TX is already finalized at this point. To figure out what you're meant to be contributing for each input you construct a plan from the descriptor + existing signatures and see what's left to do. Of course, assuming you're using PBSTs you could just blindly follow what's missing from the PSBTs but this allows some analysis before doing so e.g. construct a plan with only the signatures + what you have locally and see if it's possible for you alone to finish the PSBT before actually engaging signing devices.

I think it would be a good idea to get rid of the signatures themselves from this struct though. If they've already signed then those keys can be treated as an asset. Just map those signatures to their asset keys but there's no need to actually represent them like that in the "assets" I think.

Reply:
Oh right after writing this I realized that it's possible the signatures are only valid for one tapleaf but not the other. What if we generalized this to conditioning public keys that can only sign on particular tap leaves? E.g. maybe I have a hardware device that will only sign on this tapleaf one normally but might sign on this tap leaf (or do tap key) if some condition is met. Then existing signatures would map to keys that can only sign on a particular leaf (or as tap key).

Reply:
>What if we generalized this to conditioning public keys that can only sign on particular tap leaves?

Yes, I agree that we should do this. Otherwise I don't really see the point of supporting signatures in lieu of public keys.

Reply:
As of 4849893 this is resolved.